# Apriori Association Rule on TV Shows Dataset

Import the required libraries

In [1]:
import pandas as pd
from apyori import apriori
import numpy as np

Read the dataset and display it

In [2]:
data_set = pd.read_csv("dataset\TV Shows - Association Rule Learning.csv", header=None)
data_set

0              1                   2                 3   \
0         Cobra Kai          Lupin          12 Monkeys          Sherlock   
1              Lost      Jack Ryan           The Flash   Game of thrones   
2     Sex Education      Dr. House             Kingdom  The Walking Dead   
3             Ozark  Sex Education         Constantine          Preacher   
4            Naruto            NaN                 NaN               NaN   
...             ...            ...                 ...               ...   
9685      One Piece  The Blacklist  Two and a half men             Lupin   
9686      One Piece      Mr. Robot          Succession             Ozark   
9687      Daredevil       Atypical               Heros               NaN   
9688       Absentia   The Newsroom        The Alienist               NaN   
9689       Atypical            NaN                 NaN               NaN   

                  4                      5                    6       7   \
0                NaN                    NaN                  NaN     NaN   
1     House of Cards             12 Monkeys              Vikings  Fringe   
2                NaN                    NaN                  NaN     NaN   
3            Vikings               The Tick                  NaN     NaN   
4                NaN                    NaN                  NaN     NaN   
...              ...                    ...                  ...     ...   
9685            Dark  How I met your mother                  NaN     NaN   
9686      12 Monkeys                Vikings  The Vampire Diaries     NaN   
9687             NaN                    NaN                  NaN     NaN   
9688             NaN                    NaN                  NaN     NaN   
9689             NaN                    NaN                  NaN     NaN   

                 8             9   ...   22   23   24   25   26   27   28  \
0               NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1     The Mentalist  The Alienist  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2               NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
3               NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4               NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...             ...           ...  ...  ...  ...  ...  ...  ...  ...  ...   
9685            NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9686            NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9687            NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9688            NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9689            NaN           NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       29   30   31  
0     NaN  NaN  NaN  
1     NaN  NaN  NaN  
2     NaN  NaN  NaN  
3     NaN  NaN  NaN  
4     NaN  NaN  NaN  
...   ...  ...  ...  
9685  NaN  NaN  NaN  
9686  NaN  NaN  NaN  
9687  NaN  NaN  NaN  
9688  NaN  NaN  NaN  
9689  NaN  NaN  NaN  

[9690 rows x 32 columns]

> We Notice that The Dataset Made out of 9690 Viewers data for 32 show daily

Now we should transform the dataframe to list of lists for apriori

In [38]:
transactions = []
for i in range(0, data_set.shape[0]):
    temp = []
    for j in range(0, data_set.shape[1]):
        if not data_set.loc[i,j] is np.nan:
            temp.append(data_set.loc[i,j])
    transactions.append(temp)

We chose the `min support` and `min confidence` based on Orange Tests

In [46]:
rules = apriori(transactions= transactions, min_support=0.005, min_confidence=0.6, min_lift=2)
association_results = list(rules)
print("The Total Rules: ", len(association_results))

The Total Rules:  21


> Notice we have 143 Rules

Now Time to display the Resuts

In [47]:
for i in range(len(association_results)):
    print(str(association_results[i]) + "\n\n")

RelationRecord(items=frozenset({'12 Monkeys', 'Dr. House', 'Sex Education'}), support=0.005159958720330237, ordered_statistics=[OrderedStatistic(items_base=frozenset({'12 Monkeys', 'Dr. House'}), items_add=frozenset({'Sex Education'}), confidence=0.6172839506172839, lift=2.4148088338641425)])


RelationRecord(items=frozenset({'Atypical', 'Berlin Station', 'Sex Education'}), support=0.009391124871001032, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Atypical', 'Berlin Station'}), items_add=frozenset({'Sex Education'}), confidence=0.6408450704225352, lift=2.5069797062552954)])


RelationRecord(items=frozenset({'Dr. House', 'Berlin Station', 'Sex Education'}), support=0.005985552115583075, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Dr. House', 'Berlin Station'}), items_add=frozenset({'Sex Education'}), confidence=0.6170212765957447, lift=2.4137812556369664)])


RelationRecord(items=frozenset({'Berlin Station', 'Sex Education', 'Family Guy'}), support=0.00681